packages/compass-connections/src/connections-manager.ts

```diff
@@ -0,0 +1,242 @@
+import { EventEmitter } from 'events';
+import { createConnectionAttempt } from 'mongodb-data-service';
+
+import type { LoggerAndTelemetry } from '@mongodb-js/compass-logging';
+import type { ConnectionInfo } from '@mongodb-js/connection-info';
+import type {
+  ConnectionAttempt,
+  DataService,
+  connect,
+} from 'mongodb-data-service';
+
+type ConnectFn = typeof connect;
+type ConnectionInfoId = ConnectionInfo['id'];
+
+export enum ConnectionsManagerEvents {
+  ConnectionAttemptStarted = 'connection-attempt-started',
+  ConnectionAttemptCancelled = 'connection-attempt-cancelled',
+  ConnectionAttemptSuccessful = 'connection-attempt-successful',
+  ConnectionAttemptFailed = 'connection-attempt-failed',
+  ConnectionDisconnected = 'connection-disconnected',
+}
+
+type ConnectionManagerEventListeners = {
+  [ConnectionsManagerEvents.ConnectionAttemptStarted]: (
+    connectionInfoId: ConnectionInfoId
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionAttemptCancelled]: (
+    connectionInfoId: ConnectionInfoId
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionAttemptSuccessful]: (
+    connectionInfoId: ConnectionInfoId,
+    dataService: DataService
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionAttemptFailed]: (
+    connectionInfoId: ConnectionInfoId,
+    error: any
+  ) => void;
+  [ConnectionsManagerEvents.ConnectionDisconnected]: (
+    connectionInfoId: ConnectionInfoId
+  ) => void;
+};
+
+export enum ConnectionStatus {
+  Disconnected = 'disconnected',
+  Connecting = 'connecting',
+  Connected = 'connected',
+  Failed = 'failed',
+}
+
+type ConnectionStatusTransitions = {
+  [event in ConnectionsManagerEvents]?: {
+    [status in ConnectionStatus]?: ConnectionStatus;
+  };
+};
+
+const connectionStatusTransitions: ConnectionStatusTransitions = {
+  [ConnectionsManagerEvents.ConnectionAttemptStarted]: {
+    [ConnectionStatus.Disconnected]: ConnectionStatus.Connecting,
+    [ConnectionStatus.Failed]: ConnectionStatus.Connecting,
+  },
+  [ConnectionsManagerEvents.ConnectionAttemptCancelled]: {
+    [ConnectionStatus.Connecting]: ConnectionStatus.Disconnected,
+  },
+  [ConnectionsManagerEvents.ConnectionAttemptFailed]: {
+    [ConnectionStatus.Connecting]: ConnectionStatus.Failed,
+  },
+  [ConnectionsManagerEvents.ConnectionAttemptSuccessful]: {
+    [ConnectionStatus.Connecting]: ConnectionStatus.Connected,
+  },
+  [ConnectionsManagerEvents.ConnectionDisconnected]: {
+    [ConnectionStatus.Connected]: ConnectionStatus.Disconnected,
+  },
+};
+
+export class ConnectionsManager extends EventEmitter {
+  private connectionAttempts = new Map<ConnectionInfoId, ConnectionAttempt>();
+  private connectionStatuses = new Map<ConnectionInfoId, ConnectionStatus>();
+  private dataServices = new Map<ConnectionInfoId, DataService>();
+
+  constructor(
+    private readonly __TEST_CONNECT_FN?: ConnectFn,
+    private readonly logger?: LoggerAndTelemetry['log']['unbound']
```

Comment:
Logger should be required, not optional. When do we not want our logging to work?

Also just from the perspective of argument positioning: an extremely optional, never to be used outside of testing `__TEST_CONNECT_FN` should probably be last, not first argument

Reply:
I made it optional because it won't make sense for CompassWeb, right? Will fix the argument ordering as well, thank you for pointing out :D

Reply:
We will have a logging and telemetry service in the web environment

Reply:
It is not yet implemented right? Meanwhile I guess I could just use the `createNoopLogger` over there?

Reply:
Also addressed in my recent update.